File: storage/db-core/src/main/java/com/coffee/gu/cart/CartItemRepositoryImpl.java
package com.coffee.gu.cart;

import com.coffee.gu.Principal;
import com.coffee.gu.enums.EntityStatus;
import com.querydsl.jpa.impl.JPAQueryFactory;
import org.springframework.stereotype.Repository;

import java.util.List;
import java.util.Optional;

import static com.coffee.gu.cart.QCartItemEntity.*;

@Repository
public class CartItemRepositoryImpl implements CartItemRepository{

    private final JPAQueryFactory jpaQueryFactory;
    private final CartItemJpaRepository cartItemJpaRepository;

    public CartItemRepositoryImpl(JPAQueryFactory jpaQueryFactory, CartItemJpaRepository cartItemJpaRepository) {
        this.jpaQueryFactory = jpaQueryFactory;
        this.cartItemJpaRepository = cartItemJpaRepository;
    }

    @Override
    public List<CartItem> findByPrincipalKey(String principalKey) {
        return jpaQueryFactory
                .selectFrom(cartItemEntity)
                .where(
                        cartItemEntity.principalKey.eq(principalKey),
                        cartItemEntity.entityStatus.eq(EntityStatus.ACTIVE)
                )
                .fetch()
                .stream()
                .map(CartItemEntity::toModel)
                .toList();
    }

    @Override
    public Optional<CartItem> findByPrincipalIncludingDeleted(String principalKey) {
        return jpaQueryFactory
                .selectFrom(cartItemEntity)
                .where(
                        cartItemEntity.principalKey.eq(principalKey)
                )
                .fetch()
                .stream()
                .map(CartItemEntity::toModel)
                .toList()
                .stream()
                .findFirst();
    }

    @Override
    public CartItem save(CartItem cartItem, Principal principal) {
        return cartItemJpaRepository.save(CartItemEntity.of(cartItem, principal)).toModel();
    }
}

File: storage/db-core/src/main/kotlin/com/coffee/gu/cart/CartItemRepositoryImpl.kt
package com.coffee.gu.cart

import com.coffee.gu.Principal
import com.coffee.gu.cart.QCartItemEntity.cartItemEntity
import com.coffee.gu.enums.EntityStatus
import com.coffee.gu.menu.MenuRepository
import com.querydsl.jpa.impl.JPAQueryFactory
import org.springframework.stereotype.Repository

@Repository
class CartItemRepositoryImpl(
    private val cartItemJpaRepository: CartItemJpaRepository,
    private val jpaQueryFactory: JPAQueryFactory,
    private val menuRepository: MenuRepository,
) : CartItemRepository {

    override fun findByPrincipalKey(principalKey: String): List<CartItem> {
        return jpaQueryFactory.selectFrom(cartItemEntity)
            .where(
                cartItemEntity.principalKey.eq(principalKey),
                cartItemEntity.entityStatus.eq(EntityStatus.ACTIVE),
            )
            .fetch()
            .map { it.toModel(menuRepository.findById(it.menuId)) }
    }

    override fun findByPrincipalKeyAndMenuIdIncludingDeleted(principalKey: String, menuId: Long): CartItem? {
        val cartItem = jpaQueryFactory.selectFrom(cartItemEntity)
            .where(cartItemEntity.principalKey.eq(principalKey), cartItemEntity.menuId.eq(menuId))
            .fetchFirst()

        return cartItem?.toModel(menuRepository.findById(cartItem.menuId))
    }

    override fun save(cartItem: CartItem, principal: Principal): CartItem {
        val entity = CartItemEntity.of(cartItem, principal)
        return cartItemJpaRepository.save(entity).toModel(menuRepository.findById(entity.menuId))
    }
}
